### PR TITLE
Migrate to typescript

### PR DESCRIPTION
**Pull request by vargasjona**
This PR has the following changes:
1) Migration from javascript to typescript
2) Add Rasa types
3) Change uuid4 to secure library
4) Expose react-native-gifted-chat components from react-native-rasa

Screen Recording 2021-12-19 at 11 04 23 PM

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -5,5 +5,6 @@
 import {AppRegistry} from 'react-native';
 import App from './App';
 import {name as appName} from './app.json';
+import 'react-native-get-random-values'
```
Comment (hungvu193) [inline index.js:8]: Do we really need this library?

File: RNRasa.tsx [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+  userName?: string;
+  botName?: string;
+  botAvatar?: string;
+}
+
+const RasaChat: FC<IRasaChat> = (props: IRasaChat) => {
+  const {
+    host,
+    onSendMessFailed = (error) => console.log('Send mess failed', error),
+    onEmptyResponse = () => console.log('The bot response is empty'),
```
Comment (hungvu193) [inline RNRasa.tsx]: I think we need to remove this console log or create a common function to log error for user.

File: RNRasa.tsx [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+      user={userData}
+      onSend={(mess) => onSend(mess)}
+      messages={messages}
+      onQuickReply={onQuickReply}
+      {...giftedChatProp}
+    />
+  );
+};
+
+export { RasaChat, Actions, Avatar, Bubble, SystemMessage, MessageImage, MessageText, Composer, Day, InputToolbar, LoadEarlier, Message, MessageContainer, Send, Time, GiftedAvatar, utils };
```
Comment (hungvu193) [inline RNRasa.tsx]: Please make sure that we will keep RNRasa as `default export`, we don't want to make user confused whenever they updated new version.

File: index.js (hunk 1 above)
Comment (hungvu193) [reply index.js:8]: I created a function called `uuidv4`, so I think we don't need this library.

File: utils.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+
+export const isValidNotEmptyArray = (array: any[]) => {
+  if (
+    typeof array !== 'undefined'
+    && array != null
+    && array.length != null
+    && array.length > 0
+  ) {
+    return true;
```
Comment (hungvu193) [inline utils.ts]: should we rewrite it to :
```typescript
 return !!(array && array.length)
```

File: utils.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+export const fetchOptions = {
+  method: 'POST',
+  headers: {
+    'Content-Type': 'application/json',
+  },
+};
+
+
```
Comment (hungvu193) [inline utils.ts]: I would be appreciated if you could fill the description for this function and other functions too.
For example:
```
/**
 * Return true or false
 * @param {Array} array Blahblah
 */
```

**Comment by hungvu193** [comment]
Awesome @jonra1993 , I left some comments, let have a look and discuss about them.
You did a great job 👍

**Comment by vargasjona** [comment]
> Awesome @jonra1993 , I left some comments, let have a look and discuss about them. You did a great job 👍

Hello @hungvu193 I am not able to find the comments you left in which files did you added?

**Review by hungvu193 — COMMENTED** [review]
You can find it here @jonra1993

File: index.js (hunk 1 above)
Comment (vargasjona) [reply index.js:8]: Hello @hungvu193 using default random javascript lib is not secure for creating uuidv4 as you can check here https://rules.sonarsource.com/javascript/tag/cwe/RSPEC-2245 and react-native-get-random-values package is needed in order that uuidv4 package can work correctly on react native apps. Random triggers errors when using a static code analysis in my opinion using uuidv4 and react-native-get-random-values lib is more secure. react-native-gifted also use it https://github.com/FaridSafi/react-native-gifted-chat/blob/master/package.json

File: RNRasa.tsx (hunk 2 above)
Comment (vargasjona) [reply RNRasa.tsx]: I agree with you

File: RNRasa.tsx (hunk 3 above)
Comment (vargasjona) [reply RNRasa.tsx]: I agree with you

File: utils.ts (hunk 4 above)
Comment (vargasjona) [reply utils.ts]: ok

File: utils.ts (hunk 5 above)
Comment (vargasjona) [reply utils.ts]: I am going to do that

File: utils.ts (hunk 4 above)
Comment (vargasjona) [reply utils.ts]: return !!(array && array?.length && array?.length > 0)
  
  I have added a length greater than 0 also in order to validate that response is not an empty array

Comment (hungvu193) [reply utils.ts]: I think `!!(array && array.length)` is ok for that case :D

File: index.js (hunk 1 above)
Comment (hungvu193) [reply index.js:8]: Cool, I agree

**Comment by hungvu193** [comment]
All good @jonra1993, I will release this version (and license to MIT) after new year vacation 👍

**Review by hungvu193 — APPROVED** [review]

**Pull request merged**